feat: add pre-commit to check we don't have real passwords in yml files

Chickensoupwithrice opened the pull request:
Dirty lil script that has the ability to check changed yaml files for keys containing "password" against a set of "allowed" default passwords. Any changed yaml file that contains a field which contains "password" and has a value that isn't in `ALLOWED_PASSWORDS` will cause the pre-commit to fail. 

Resolves https://github.com/webrecorder/browsertrix-cloud/issues/988

I tried to make it as understandable as possible, but it's still sorta ugly. I could write tests for this, but I thought it was small enough, and far enough away from the main app that it wasn't strictly necessary (Feel free to tell me otherwise!). I've also included a script that will symbolic link the correct file into the expected location for git pre-commits. 

If you like I can add `black` / `pylint` / `flake8` whatever else to the pre-commit, but for now it only handles commiting passwords in yml / yaml files in fields that contain "password", seeing as we have CI configured for the others. 

And because every good PR deserves a screenshot, here's some of my ad-hoc testing:
![image](https://github.com/webrecorder/browsertrix-cloud/assets/22575913/f0f70447-801f-432b-9df0-7eac19e398b5)

Chickensoupwithrice committed:
feat: add pre-commit to check we don't have real passwords in yml files

tw4l reviewed (commented):
Thanks @Chickensoupwithrice, this is a nice addition! We have a pre-commit config already in the main directory to run black and linters for the backend and linters via husky for the frontend - would it be possible to integrate your check_passwords script into the existing pre-commit infrastructure?

Would be great to also add this script to GH Actions to flag if a password accidentally gets committed so we can quickly remedy (e.g. in case someone doesn't install pre-commit locally)

Chickensoupwithrice commented:
Wow, I totally missed that! 
Yes, absolutely I'll integrate with `pre-commit` and can configure a github action around this as well :)

tw4l commented:
> Wow, I totally missed that! Yes, absolutely I'll integrate with `pre-commit` and can configure a github action around this as well :)

Awesome, thank you! No worries, we have a lot of config files in this repo haha

Chickensoupwithrice committed:
feat: use pre-commit framework instead

Chickensoupwithrice commented:
Alright, I dropped the bash file and the install script and switched to using pre-commit framework and just the python script. 

Screenshot of failures:
![image](https://github.com/webrecorder/browsertrix-cloud/assets/22575913/f542866c-f00e-4c48-9d0d-7b73b6430c37)

I think the frontend tests are failing, because I don't have any of the deps installed.

Chickensoupwithrice commented:
And also added the github action to this PR. 

Testing with Act once again (expected failure): 
![image](https://github.com/webrecorder/browsertrix-cloud/assets/22575913/93487306-5f6e-4fb3-9ca2-5d6bec98233e)

Expected success:
![image](https://github.com/webrecorder/browsertrix-cloud/assets/22575913/dc12cbc1-f426-4b7d-9c34-2701a81fa4d2)

Chickensoupwithrice committed:
feat(ci): add github action for password_check

Chickensoupwithrice commented:
Though this is working now, after running into some issues with it checking github actions (since they're also yaml files) I decided to write some tests for this given it's failure would be annoying. Converted to draft until tests are complete.

Chickensoupwithrice commented:
Still ugly as, but now featuring tests!

ikreymer commented:
Looks good, though I think it currently fails on our values.yaml file:
```
python ./scripts/check_passwords.py chart/values.yaml 
top level key 'backend_password_secret' in chart/values.yaml contains a real password!
top level key 'mongo_auth' with subkey 'password' in chart/values.yaml contains a real password!
```
That's a reminder that we should rotate the backend_password_secret -- could set it to one of our default passwords, and also set the mongo_auth `PASSWORD!` as not a real password..

Chickensoupwithrice committed:
feat: add some simple tests to password_check.py

Chickensoupwithrice committed:
fix: set `backend_password_secret` to an allowed password

Chickensoupwithrice commented:
I added "PASSWORD!" to the allowed passwords list, and also set backend_password_secret to it. I didn't rotate the password in our actual deployment though.
